### PR TITLE
Prepare for fixing packing of Rates, fix bug in Connection comparison operator

### PR DESCRIPTION
**Pull request by atgeirr**
After adding the `tracer` member the data::Rates class, it is no longer a POD. The same applies to data::Connection and data::Segment, since these contain data::Rates subobjects. In the serialization used for communicating these objects to other MPI ranks, this means they now need specific packSize(), pack() and unpack() methods. These must be implemented in opm-simulators (in ParallelRestart.cpp), but need access to the members. I have therefore added a friend declaration that enables this.

The second commit is a drive-by fix in the same file, operator== of Connection was not updated when adding the `trans_factor` member. This fix then caused a test failure, it turns out that the restart I/O does not treat the Connection objects completely. However, the data::Wells objects compared are never used in actual restarting, so it is not something we need to test! (Thanks to @bska for explanation.) I have therefore removed these comparisons from the test, this is the third commit of the PR.

A final note on serialization. We **really** should implement unified serialization in-class using the serializeOp approach. That would result in less code and fewer bugs.

### CODE REVIEW

**Comment by atgeirr** [comment]
I should note that a downstream PR in opm-simulator will come that depends on this and fixes restart in parallel. This should be mergeable on its own though.

**Comment by atgeirr** [comment]
jenkins build this please

File: opm/output/data/Wells.hpp
Comment (bska) [inline opm/output/data/Wells.hpp:106]: So this will, in principle, mean that `Mpi::Packing<MyEvilType>` has read/write access to the private data members of `data::Rates`.  It probably doesn't matter, but is something we should keep in mind.  Is there a way to limit **which** specializations of `Mpi::Packing<>` are granted special access privileges?  Maybe something along the lines of
```C++
friend struct Mpi::Packing<Rates>
```

**Review by bska — COMMENTED** [review]
Good catch and I guess we were just unlucky that this issue did not present earlier.  Hooking this up to the `serializeOp()` machinery is probably a good long-term goal, but I think we'll have to continue to support the `read()`/`write()` protocol too.  At least until/unless we massively overhaul the [`CollectDataToIORank`](https://github.com/OPM/opm-simulators/blob/6d3d64a8437b31eb06763e97cbb2aa273b2cfcc3/ebos/collecttoiorank.hh#L47) facility in opm-simulators.

I have just one concern about the rather permissive access privileges being granted here.  I don't think it's a show-stopper, but I wondered if there's a way to make it a bit more restrictive.

File: opm/output/data/Wells.hpp
Comment (atgeirr) [reply opm/output/data/Wells.hpp:106]: I will check if that works.

Comment (atgeirr) [reply opm/output/data/Wells.hpp]: It works, and I have changed it.

**Comment by atgeirr** [comment]
jenkins build this please

**Review by bska — APPROVED** [review]
Looks good to me now, and I'm prepared to merge this.  On the other hand, the [downstream test failures](https://ci.opm-project.org/job/opm-simulators-PR-builder/3886/testReport/) suggest we **might** need the same kind of treatment for the [`SegmentPressures`](https://github.com/OPM/opm-common/blob/70d45e74dce5fb631a193d53597de47fb16a8392/opm/output/data/Wells.hpp#L173) too.  We might also just be missing a `HANDLE_AS_POD` treatment for those...

Would you like to add the `Mpi::Packing<SegmentPressures>` friend declaration here before we merge this?

**Comment by atgeirr** [comment]
> Would you like to add the `Mpi::Packing<SegmentPressures>` friend declaration here before we merge this?

Yes I will. It is not correct to add a HANDLE_AS_POD thing, it is SegmentPressures not being a POD that triggered the test failure.

**Comment by atgeirr** [comment]
jenkins build this please

**Comment by atgeirr** [comment]
All green and approved, merging.

**Pull request merged**
